### PR TITLE
SW-3696 Services & redux for observations in the FE

### PR DESCRIPTION
**Pull request by karthikbtf**
- this is off Steve's [BE PR](https://github.com/terraware/terraware-server/pull/1183), the FE PR will not build in staging
- generated schema and added types/models
- updated existing code to match some breaking model changes on planting sites
- added a service layer that makes one call for now (fetch full hierarhical observations results data)
- tested locally on an org that has observations (generated using Steve's BE script)
- see [gist](https://gist.github.com/karthikbtf/e6a282831ab94414fe7af14eed2a1d81) of what end results looks like in my org
- we may need more services/redux on top of the one API in the service, those will not be covered in this PR.. the current single API covers all the observations page-view requirements, plants dashboard and observation settings related API will need to be separate PRs when the time comes
- created Redux layers
- created initial view of observations using placeholder content
- have observations page fetch real data to decide when to show empty content vs data view

Terraware App 2023-06-02 14-39-03

### CODE REVIEW

**Review by charlie-tf — COMMENTED** [review]
Looks good so far

**Comment by karthikbtf** [comment]
> Looks good so far

Thanks, will be working on the redux layer next. Going to start simple here and won't add any utilities etc. We can evolve them as we see fit.

**Comment by karthikbtf** [comment]
@charlie-tf you could if you wish,, branch off this and create the map component for observations.
You would have to use steve's branch for the BE locally.

File: src/components/Observations/ObservationsView.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  selectCompletedObservationsResults,
+  selectPlantingSiteObservationsResults,
+} from 'src/redux/features/observations/observationsSelectors';
+import ListMapView from 'src/components/ListMapView';
+
+export type ObservationsViewProps = {
+  selectedPlantingSiteId: number;
+};
+
+export default function ObservationsView({ selectedPlantingSiteId }: ObservationsViewProps): JSX.Element {
```
Comment (karthikbtf) [inline src/components/Observations/ObservationsView.tsx:12]: I will rename this to ObservationsDataView - that's the representation here

File: src/components/Observations/index.tsx [hunk 2]
```diff
@@ -41,10 +67,12 @@ export default function PlantsDashboard(): JSX.Element {
...
       allowAllAsSiteSelection={true}
       onPlantingSites={onPlantingSites}
-      isEmptyState={!plantingSites?.length || !hasObservations}
+      isEmptyState={!plantingSites?.length || !observationsResults?.length}
     >
-      {hasObservations ? (
-        <div>placeholder for selected planting site {selectedPlantingSite?.id}</div>
+      {observationsResults === undefined && observationsResultsError === undefined ? (
+        <CircularProgress sx={{ margin: 'auto' }} />
+      ) : observationsResults?.length && selectedPlantingSite ? (
```
Comment (karthikbtf) [inline src/components/Observations/index.tsx:74]: i have to update this,, show empty state if a specific selected planting site has no observations.

File: src/redux/features/observations/observationsSelectors.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { RootState } from 'src/redux/rootReducer';
+import { createSelector } from '@reduxjs/toolkit';
+
+export const selectObservationsResults = (state: RootState) => state.observationsResults?.observations;
+export const selectObservationsResultsError = (state: RootState) => state.observationsResults?.error;
```
Comment (charlie-tf) [inline src/redux/features/observations/observationsSelectors.ts:5]: `state.observationResults` comes from `name: 'observationResults'` in the slice file?

File: src/redux/features/observations/observationsSelectors.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+export const selectCompletedObservationsResults = createSelector(selectObservationsResults, (observationsResults) =>
+  observationsResults?.filter((observationResults) => !!observationResults.completedTime)
+);
+
+export const selectPlantingSiteObservationsResults = createSelector(
+  [selectCompletedObservationsResults, (state, plantingSiteId) => plantingSiteId],
```
Comment (charlie-tf) [inline src/redux/features/observations/observationsSelectors.ts:12]: What does this list do?

File: src/redux/rootReducer.ts [hunk 5]
```diff
@@ -1,9 +1,11 @@
 import { Action, combineReducers } from '@reduxjs/toolkit';
 import { appVersionReducer } from './features/appVersion/appVersionSlice';
+import { observationsResultsReducer } from './features/observations/observationsSlice';
 
 // assembly of app reducers
 export const reducers = {
   appVersion: appVersionReducer,
+  observationsResults: observationsResultsReducer,
```
Comment (karthikbtf) [inline src/redux/rootReducer.ts:8]: @charlie-tf the `state.observationResults` comes from this key

File: src/redux/features/observations/observationsSelectors.ts (hunk 4 above)
Comment (karthikbtf) [reply src/redux/features/observations/observationsSelectors.ts:12]: The observations page has a planting site selector. This list returns observations for the selected planting site - including 'All'.

Comment (karthikbtf) [reply src/redux/features/observations/observationsSelectors.ts:12]: One thing about using selectors is that we use selectors to access data. That includes narrowing down of data.

**Comment by karthikbtf** [comment]
> This is looking good. Just a couple minor questions about the setting up of the selectors

Thanks. I will rename the slice to avoid confusion, and update the completedTime to `yyyy-mm-dd` format with planting site time zone accounted for (so downstream code needn't worry about it).

File: src/redux/rootReducer.ts (hunk 5 above)
Comment (charlie-tf) [reply src/redux/rootReducer.ts:8]: Ohh, gotcha 👍

File: src/redux/features/observations/observationsSelectors.ts (hunk 4 above)
Comment (charlie-tf) [reply src/redux/features/observations/observationsSelectors.ts:12]: Got it, Thanks

**Comment by karthikbtf** [comment]
> > This is looking good. Just a couple minor questions about the setting up of the selectors
> 
> Thanks. I will rename the slice to avoid confusion, and update the completedTime to `yyyy-mm-dd` format with planting site time zone accounted for (so downstream code needn't worry about it).

Updated PR to map all completed times to planting site time zone. Also renamed the slice.
If the BE PR has any updates that require changes here, I will follow up (won't merge until then ofc).

Terraware App 2023-06-05 11-17-15

**Review by charlie-tf — APPROVED** [review]
Changes look good 👍

File: src/redux/features/observations/observationsThunks.ts [hunk 6]
```diff
@@ -1,14 +1,63 @@
...
       const response = await ObservationsService.listObservationsResults(organizationId);
       const { error, observations } = response;
-      dispatch(setObservationsResultsAction({ error, observations }));
+      dispatch(
+        setObservationsResultsAction({
+          error,
+          observations: observations.map((observation) =>
+            applyTimeZone(
+              observation,
+              plantingSiteTimeZones.find((site) => observation.plantingSiteId === site.id)?.timeZone
```
[the remark this answers is not in the document]
Comment (karthikbtf) [reply src/redux/features/observations/observationsThunks.ts:56]: The timezone?

[the remark this answers is not in the document]
Comment (charlie-tf) [reply src/redux/features/observations/observationsThunks.ts:56]: Yes, that's what I meant, sorry. Will there be a case in which it cannot find a timezone for the plantingSite. I was wondering how the applyTimeZone method handled the case where it was undefined. Digging into it a little, I see that it defaults to `Etc/UTC`. No changes or anything needed, I was just curious

**Review by charlie-tf — APPROVED** [review]

**Pull request merged**